pyro/infer/reparam/neutra.py

```diff
@@ -0,0 +1,40 @@
+import pyro
+import pyro.distributions as dist
+from pyro.infer.autoguide.guides import AutoContinuous
+
+
+class NeuTraReparam:
+    """
+    Example::
+
+        neutra = NeuTraReparam(AutoIAFNormal(model))
+        model = poutine.infer_config(lambda _: {"reparam": neutra})
+        model = poutine.reparam(model)
+        # ...now use the model in HMC or NUTS...
+
+    Note that all sites must share a common ``NeuTra`` instance.
+
+    [1] Hoffman, M. et al. (2019)
+        "NeuTra-lizing Bad Geometry in Hamiltonian Monte Carlo Using Neural Transport"
+        https://arxiv.org/abs/1903.03704
+    """
+    def __init__(self, guide):
+        assert isinstance(guide, AutoContinuous)
+        self.guide
+        self.x_constrained = []
+
+    def __call__(self, name, fn, obs):
+        log_density = 0.
+        if not self.x_constrained:
+            # Sample a shared latent on first sample site.
+            posterior = self.guide.get_posterior()
+            x_unconstrained = pyro.sample("{}_latent".format(name), posterior)
+            log_density = -posterior.log_prob(x_unconstrained)
+            x_constrained = self.guide._unpack_latent(x_unconstrained)
+            self.x_constrained = list(reversed(x_constrained))
+
+        # Extract a single site's value from the shared latent.
+        site, value = self.x_unconstrained.pop()
+        log_density = log_density + fn.log_prob(value)
```

Comment:
Thanks for explaining, I think I understand now. Does this look correct now?

Reply:
I think that you are reparameterized the `guide` here instead of the `model`. IIUC, then we need to (assume that `fn` is latent of `model`) leverage the logic [here](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/autoguide/guides.py#L458).
```
site, unconstrained_value = self.x_unconstrained.pop()
transform = biject_to(fn.support)
value = transform(unconstrained_value)
logdet = transform.inv.log_abs_det_jacobian(value, unconstrained_value)
logdet = sum_rightmost(logdet, logdet.dim() - value.dim() + fn.event_dim)
log_density = log_density + fn.log_prob(value) + logdet
```

Reply:
done